### PR TITLE
feat: stochastic Allen-Cahn SPDE stepper and supporting infrastructure

### PR DESCRIPTION
**Pull request by GVourvachakis**
# feat: add stochastic Allen-Cahn SPDE stepper and supporting infrastructure

This commit introduces first-class support for stochastic PDEs (SPDEs)
in `exponax`, structured to allow incremental addition of further SPDEs
(e.g. Kardar-Parisi-Zhang, stochastic Navier-Stokes) without
excessive intervention in the source code.

---

## 1. `exponax/stepper/stochastic/` — new subpackage

Adds a `stochastic/` namespace inside the existing `stepper/` hierarchy,
mirroring the pattern of `stepper/reaction/`, `stepper/convection/`, etc.

### `__init__.py`
Public API: exports the `stochastic/` subpackage.
Module-level docstring documents the subpackage purpose,
lists known limitations with references,
and explains the ETD operator-splitting convention so future SPDE
authors can follow the same pattern.

### `_stochastic_allen_cahn.py`
Implements `StochasticAllenCahn`, a $d$-dimensional (d = 1, 2, 3)
SPDE stepper on periodic domains using the Exponential
Euler-Maruyama (EEM) method (Lord, Powell & Shardlow, 2014, Chapters 7–10).

The equation solved (Itô form) is:

$$\partial_t u = \nu \Delta u + \lambda(u - u^3) + \sigma(u)\,\xi(x, t)$$

where $\xi$ is a Q-Wiener process with spectral covariance
$Q_k \propto (1 + |k|^2_{\rm phys})^{-\alpha}$.

**Key design decisions:**

- The ETD operator splitting follows the built-in `reaction.AllenCahn`
  convention exactly ($L = \nu\Delta + \lambda$, $\mathcal{N} = -\lambda u^3$),
  so $\sigma = 0$ reproduces the deterministic stepper to machine precision.
- Exact modewise stochastic-integral variance is used:

  $$\mathrm{Var}_k = Q_k \cdot \left(e^{2L_k \Delta t} - 1 \right) / \left( 2L_k \right)$$

  which avoids the $O(\Delta t)$ approximation of naive Euler-Maruyama.
- Both additive ($\sigma(u)=\sigma$) and multiplicative ($\sigma(u)=\sigma u$)
  noise are supported; $\sigma$ is embedded in a precomputed `_noise_std`
  array to prevent double-counting bugs.
- Optional Hutzenthaler-Jentzen taming (`use_taming=True`) of the cubic
  nonlinearity provides almost-sure boundedness for large $\lambda$ or
  $\sigma$ (Hutzenthaler & Jentzen, 2015).
- Optional first-order Milstein correction for multiplicative noise
  (`use_milstein=True`); auto-disabled for $\alpha = 0$ (space-time white
  noise) where the iterated stochastic integrals diverge.
- `BaseStepper.step()` is deliberately disabled and raises
  `NotImplementedError`; callers must supply a PRNGKey via
  `stepper(u, key=key)` or `stepper.step_fourier(u_hat, key=key)`.
- `step_fourier` avoids the round-trip to physical space for callers
  that already hold the Fourier representation.
- Three documented known limitations (DC/Nyquist variance halving,
  non-standard Milstein ETD prefactor, noise above dealiasing cutoff)
  are explained with literature references.

---

## 2. `exponax/nonlin_fun/_tamed_polynomial.py` — new helper

Adds `TamedPolynomialNonlinearFun`, a drop-in replacement for the
existing `PolynomialNonlinearFun` that adds the optional taming
denominator from Hutzenthaler & Jentzen (2015):

$$\mathcal{N}_{\rm tamed}(u) = \frac{\mathcal{N}(u)}{1 + \Delta t \ |\mathcal{N}(u)|}$$

The helper inherits from `BaseNonlinearFun` and follows the exact same
interface as `PolynomialNonlinearFun`, so it is a zero-friction drop-in
for any future SPDE stepper whose nonlinear term grows super-linearly
(polynomial order $\geq 3$). `StochasticAllenCahn` uses this class for
its cubic $-\lambda u^3$ term. `use_taming=False` recovers standard
polynomial evaluation with no runtime cost.

`exponax/nonlin_fun/__init__.py` is updated to export
`TamedPolynomialNonlinearFun` alongside the existing exports.

---

## 3. `exponax/_stochastic_utils.py` — stochastic utilities module

Provides five public functions, all exported from `exponax/__init__.py`
for direct use as `ex.<function>`:

| Function | Description |
|----------|-------------|
| `stochastic_rollout(stepper, T, *, include_init)` | Thin wrapper around `jax.lax.scan` threading a PRNGKey through $T$ steps; returns trajectory `(T+1, C, *N)`. JIT-compatible and vmappable. |
| `stochastic_ensemble_rollout(stepper, T, M, *, include_init)` | Runs $M$ independent trajectories via `jax.vmap`, returning `(M, T+1, C, *N)`. Primary workhorse for all Monte Carlo diagnostics. |
| `structure_factor(ensemble, *, burn_in_fraction)` | Ensemble- and time-averaged power spectrum $S(k)$ in coefficient-extraction scaling for direct comparison to analytical $C_k$. Uses `round=None` to avoid silently zeroing small-amplitude high-k modes. |
| `richardson_weak_extrapolation(stepper_coarse, stepper_fine, u0, num_steps_coarse, key, *, num_samples)` | Runs ensembles at $\Delta t$ and $\Delta t/2$; returns Richardson-extrapolated mean, doubling weak convergence order. |
| `strang_split_step(spectral_stepper, ssa_step_fn, u, ssa_state, dt, key, *, domain_extent, num_points, mollifier_cutoff)` | Second-order Strang splitting scaffold for hybrid spectral-PDE / discrete-SSA time integration: half-SSA($\Delta t/2$) → full-PDE($\Delta t$) → half-SSA($\Delta t/2$). Band-limited Fourier mollifier suppresses Gibbs artefacts. **Not JIT-compatible across steps** when `ssa_step_fn` is a Python-level callable. References: Harrison & Yates (2016), Strang (1968), Gillespie (1977). |

---

## 4. `tests/test_stochastic/` — new SPDE test subpackage

A self-contained test directory with a `conftest.py` that registers a
`@pytest.mark.slow` marker, enabling selective execution:

```bash
# Fast tests only (~30 s on CPU):
JAX_ENABLE_X64=1 pytest tests/test_stochastic/ -m "not slow"

# Full suite (~2 mins on CPU):
JAX_ENABLE_X64=1 pytest tests/test_stochastic/
```

`test_allen_cahn.py` contains six test classes and a JAX-compatibility suite:

| Class | Marker | What is asserted |
|-------|--------|-----------------|
| `TestDeterministicLimit` | — | `sigma=0` reproduces `reaction.AllenCahn` to machine precision in 1-D and 2-D; no noise leakage; `step()` raises `NotImplementedError`. |
| `TestLinearSPDEInvariantMeasure` | slow | $\lambda=0$ (stochastic heat equation): empirical $$S(k)$$ to $$C_k$$ in 1-D (median rel-err < 0.25) and 2-D (< 0.30, below dealiasing cutoff). |
| `TestStrongConvergenceOrder` | slow | Log-log slope of mean $L^2$ error vs $\Delta t$ lies in $[0.3, 0.8]$. Wide tolerance acknowledges absence of exact path coupling. |
| `TestStructureFactorGridConvergence` | slow | $N$-normalised $S(k)/N^d$ converges between $N=64$ and $N=128$ (1-D) and $N=32$ and $N=64$ (2-D). |
| `TestMeanVarianceTimeSeries` | — | Ensemble mean $\approx 0$ (zero-mean IC, additive noise); variance positive and bounded in 1-D and 2-D. |
| `TestMilsteinCorrection` | slow | Sanity check for the Milstein correction under multiplicative noise. The non-standard $\varphi_1(L_k\Delta t)\Delta t$ ETD prefactor (Known Limitation 2) means no strict error ordering is guaranteed. The test asserts: (1) finite output (no NaN/Inf); (2) ensemble mean within $20\times$ reference field magnitude; (3) ensemble variance $\leq 10\times$ fine-Δt reference variance. The `err_eem / err_mil` ratio is printed diagnostically without being asserted. |
| `TestJAXCompatibility` | — | JIT, vmap over keys, vmap over ICs, `eqx.filter_vmap` over stacked steppers, PyTree finiteness, multiplicative noise amplitude, `step_fourier`, rollout scan. |

---

## 5. `exponax/validation/validate_stochastic_allen_cahn.ipynb`

A 14-section Jupyter notebook providing narrative theory, executable code,
printed pass/fail assertions, and a final summary table.

| Section | Topic |
|---------|-------|
| 1 | Deterministic limit: $\sigma=0$ matches `reaction.AllenCahn` in $L^\infty$ |
| 2 | 1-D invariant measure: empirical $S(k)$ vs $C_k = Q_k/(2\nu\|k\|^2)$ |
| 3 | 2-D invariant measure: structure-factor heat maps and ratio plot $S/C_k$ |
| 4 | Noise colour sweep $\alpha \in \{0,1,2,3\}$: PSD slope vs $\alpha$ |
| 5 | Additive vs multiplicative noise: variance growth and ensemble mean evolution |
| 6 | IC sensitivity: `GaussianRandomField`, `WhiteNoise`, flat IC |
| 7 | 1-D visualisation: spatio-temporal diagram and animation |
| 8 | 2-D visualisation: snapshots, animation, radial PSD, coarsening diagnostics ($k_{\rm int} = \sqrt{\lambda/\nu}$) |
| 9 | 3-D visualisation: mid-slices and volume animation via `animate_state_3d` (`vape4d` optional) |
| 10 | Strong convergence: measured slope vs $\Delta t^{0.5}$; path-coupling caveat and same-key protocol |
| 11 | Milstein vs EEM: weak-error slopes and per-step timing over two decades of $\Delta t$ ($M=2000$) |
| 12 | Richardson weak extrapolation: bias reduction using `ex.richardson_weak_extrapolation` |
| 13 | Hybrid SSA scaffold: `strang_split_step` with OU sub-step ($\theta=5$, $\mu=-0.3$) vs Allen-Cahn $+1$ attractor |
| 14 | Summary table: all measured vs expected quantities with ✓/✗ |

---

## 6. `pyproject.toml` — minor additions

- **`[tool.pytest.ini_options]`**: registers the `slow` marker globally so
  `-m "not slow"` works without per-file `conftest` duplication.
- **`dependency-groups`**: adds `pytest`, `pytest-cov`, and `ipykernel` as
  optional development dependencies (not required for library use;
  can be removed if the project prefers a leaner `pyproject.toml`).
- **`keywords`**: adds `"statistical-physics"` and `"stochastic-processes"`
  to improve PyPI discoverability.

---

### `exponax/stepper/__init__.py`
Extended module docstring now includes a **"Stochastic steppers"** section
documenting `StochasticAllenCahn`: the Itô PDE form, noise types,
Q-Wiener covariance with colour index $\alpha$, taming option, special
reductions ($\lambda=0 \to$ stochastic heat equation; $\sigma=0 \to$
deterministic), dimensional support $d = 1,2,3$, and the mandatory
key-passing calling convention.
Import chain: `exponax.stepper` now re-exports `StochasticAllenCahn`
via `from .stochastic import StochasticAllenCahn`.

### `exponax/nonlin_fun/__init__.py`
Exports `TamedPolynomialNonlinearFun` alongside the existing
`PolynomialNonlinearFun`, `BaseNonlinearFun`, and `ConvectionNonlinearFun`.

### `exponax/__init__.py`
Top-level namespace gains five new stochastic utilities:

```python
ex.stochastic_rollout
ex.stochastic_ensemble_rollout
ex.structure_factor
ex.richardson_weak_extrapolation
ex.strang_split_step
```

All are added to `__all__` and documented in the module docstring
under a new **"Stochastic utilities"** section.

---

## Known limitations and future work

The following are documented explicitly in the source and noted here
for reviewer awareness:

1. **DC and Nyquist mode variance**: independent real/imaginary Gaussians
   are drawn for all `rfft` entries including the DC ($k=0$) and Nyquist
   ($k=N/2$) modes, which must be purely real for real-valued fields.
   Taking `.real` after `ifft` silently halves their per-step variance.
   A rigorous fix would project the noise at those indices to zero.

2. **Non-standard Milstein prefactor**: the Milstein correction is scaled
   by $\varphi_1(L_k\Delta t)\Delta t$ (treating it as an ETD nonlinear
   term). The textbook EEM-Milstein scheme (Jentzen & Kloeden, 2009a)
   applies the correction without this ETD factor. The measured weak-error
   slope for Milstein is therefore closer to 1 than to 2.

3. **Noise above the dealiasing cutoff**: `_noise_std` is not masked by
   the 2/3-dealiasing filter, consistent with the Q-Wiener discretisation
   convention (Lord et al., 2014, Ch. 10) but diverging from schemes that
   band-limit the forcing.

4. **Strong convergence test**: uses a same-key proxy rather than exact
   path coupling (which would require storing and replaying fine-grid noise
   increments). The measured slope is a lower bound on the true strong order.

5. **`strang_split_step` JIT limitation**: not JIT-compatible across steps
   when the SSA sub-step uses a Python-level RNG.

6. **Scope**: only `StochasticAllenCahn` is provided. The `stochastic/`
   subpackage is explicitly designed to accept future additions such as a
   stochastic Kuramoto-Sivashinsky, Kardar-Parisi-Zhang, or stochastic
   Navier-Stokes stepper following the same pattern.

### CODE REVIEW

**Comment by Ceyron** [comment]
Hi,

Thanks a lot for your interest in Exponax and the goal to enhance it 😊. For now, I would like to keep the set of equations as is to have the package as lean as possible. I could imagine setting up an awesome list in the documentation to link to code that builds custom solvers on top of Exponax.
If you set up a nice repo that explains the motivation for your stepper and PDE and has a few examples, I am happy to link to it from the docs.

**Comment by GVourvachakis** [comment]
Hi Felix,

Thank you for taking the time to review the PR and for the kind words and completely understood. 
Keeping exponax lean and focused makes sense as a design philosophy, and I respect that.
Your suggestion of a standalone repository with a link from the docs is a great alternative. 
I will set one up that explains the motivation for the stochastic Allen-Cahn stepper, 
documents how it builds on exponax's ETD infrastructure, and includes the validation notebook and examples. 
Once it is in good shape I will come back and suggest a link for the *awesome list* you mentioned.

Thanks again for the quick response and for the pointer on how to contribute in a way that fits the project's direction.

**Comment by GVourvachakis** [comment]
Hi Felix, 

Following your suggestion I have set up a standalone repository at https://github.com/GVourvachakis/exponax-spde. It includes the `StochasticAllenCahn` EEM stepper, stochastic utilities, a 14-section validation notebook, and a guide for extending to new SPDEs. 
Happy to have it linked from the `exponax` docs whenever convenient.

Let me know your thoughts on the logo.
Three main extensions on your svg logo:
- A growing spread ∝ √t and all 8 paths start at a shared point (the shared `u0`) and fan out proportionally as `x` increases (translated Wiener process).
- A faint Gaussian envelope at the right edge showing `p(u, T)`, essentially what the `structure_factor()` computes (marginal distribution profile).
- The two-circle marker at (350, 148) (center) represents the shared `u0` from which all trajectories are seeded, with the outer glow suggesting the initial Dirac δ measure before diffusion begins.

**Comment by GVourvachakis** [comment]
Hi Felix,

Hope you are well.
I would appreciate it if you take a look at my recent package, [exponax-spde](https://github.com/GVourvachakis/exponax-spde) which is public on GitHub; an extension of [exponax](https://github.com/Ceyron/exponax) for SPDEs.

Let me know your thoughts and recommendations. 
I have left open a lot of extensions which are easy to implement as I mention in [README.md](https://github.com/GVourvachakis/exponax-spde/blob/main/README.md) and [docs/extending.md](https://github.com/GVourvachakis/exponax-spde/blob/main/docs/extending.md).

If you appreciate this repository and think it is useful for practitioners and researchers, feel free to star it or even include it in the `exponax`'s docs (e.g., in "Getting Started", "Additional", or "Additional API" $$\to$$ extensions).

Thanks for your feedback and recommendations.

**Comment by Ceyron** [comment]
Hi,

Thanks for putting in the effort to set up the repository. To have your subpackage follow `exponax`'s developments, I think it would be best if you reused or re-exported submodules of `exponax`. At the moment, it looks like you deep-copied the repository and made changes to it. 

I will set up the awesome page soon.

**Comment by GVourvachakis** [comment]
Hi again,
The hardcopy was done to ensure a more lightweight package, as plenty of features are not being used (e.g., [_vorticity_convection.py](https://github.com/Ceyron/exponax/blob/main/exponax/nonlin_fun/_vorticity_convection.py) or [_hyper_diffusion.py](https://github.com/Ceyron/exponax/blob/main/exponax/stepper/_hyper_diffusion.py)).
I removed them as they are not being used for the case of stochastic Allen-Cahn (I left Burgers for the pytest of the analytic Cole-Hopf solution).

However, since in the [pyproject.toml](https://github.com/GVourvachakis/exponax-spde/blob/main/pyproject.toml) I include the `exponax` package (at least v0.2.0), I agree that it would be better to reuse some of the utilities and helpers from the actual `exponax`.

I will include it as an [issue](https://github.com/GVourvachakis/exponax-spde/issues/3) as I work on a different project at that time, and I will update the `eponax-spde` accordingly in the next PR.
